pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java

```diff
@@ -176,76 +196,406 @@ public void shutdown() {
+
+    if (externalViewRecordVersion != lastKnownExternalViewVersion || lastKnownExternalViewVersion == INVALID_EXTERNAL_VIEW_VERSION) {
+      LOGGER.info(
+          "Routing table for table {} requires rebuild due to external view change (current version {}, last known version {})",
+          tableName, externalViewRecordVersion, lastKnownExternalViewVersion);
+      return true;
+    }
+
+    // Check if there are relevant instance config changes
+    Map<String, InstanceConfig> lastKnownInstanceConfigs = _lastKnownInstanceConfigsForTable.get(tableName);
+    if (lastKnownInstanceConfigs == null || lastKnownInstanceConfigs.isEmpty()) {
+      LOGGER.info("Routing table for table {} requires rebuild due to empty/null previous instance configs", tableName);
+      return true;
+    }
+
+    // Gather relevant incoming instance configs
+    Map<String, InstanceConfig> currentRelevantInstanceConfigs = new HashMap<>();
+    for (InstanceConfig incomingInstanceConfig : instanceConfigs) {
+      String instanceName = incomingInstanceConfig.getInstanceName();
+
+      if (lastKnownInstanceConfigs.containsKey(instanceName)) {
+        currentRelevantInstanceConfigs.put(instanceName, incomingInstanceConfig);
       }
+    }
 
+    // Did some instances lose their configuration?
+    if (lastKnownInstanceConfigs.size() != currentRelevantInstanceConfigs.size()) {
       LOGGER.info(
-          "Updating routing table for table {} due to ZK change (current version {}, last known version {})",
-          tableName, externalViewRecordVersion, lastKnownZkVersion);
+          "Routing table for table {} requires rebuild due to having a different number of instance configs (known instance config count {}, current instance config count {})",
+          tableName, lastKnownInstanceConfigs.size(), currentRelevantInstanceConfigs.size());
+      return true;
+    }
+
+    // Did some instance change state?
+    for (String instanceName : lastKnownInstanceConfigs.keySet()) {
+      InstanceConfig previousInstanceConfig = lastKnownInstanceConfigs.get(instanceName);
+      InstanceConfig currentInstanceConfig = currentRelevantInstanceConfigs.get(instanceName);
+
+      // If it's the same znode, don't bother comparing the contents of the instance configs
+      if (previousInstanceConfig.getRecord().getVersion() == currentInstanceConfig.getRecord().getVersion()) {
+        continue;
+      }
+
+      // Check if the instance got enabled/disabled or started/stopped shutting down since the last update
+      boolean wasEnabled = previousInstanceConfig.getInstanceEnabled();
+      boolean isEnabled = currentInstanceConfig.getInstanceEnabled();
+
+      String wasShuttingDown =
+          previousInstanceConfig.getRecord().getSimpleField(CommonConstants.Helix.IS_SHUTDOWN_IN_PROGRESS);
+      String isShuttingDown =
+          currentInstanceConfig.getRecord().getSimpleField(CommonConstants.Helix.IS_SHUTDOWN_IN_PROGRESS);
+
+      boolean instancesChanged =
```

Comment:
Is it, because if we're at that point in the code, the znode versions are different. For reference, the code above is:

```
      // If it's the same znode, don't bother comparing the contents of the instance configs
      if (previousInstanceConfig.getRecord().getVersion() == currentInstanceConfig.getRecord().getVersion()) {
        continue;
      }
```

Hence we can only be at that place in the code if the znodes are different.